src/Codec/Xlsx/Lens.hs

```diff
@@ -21,55 +26,97 @@ module Codec.Xlsx.Lens
 import Codec.Xlsx.Types
 #ifdef USE_MICROLENS
 import Lens.Micro
+import Lens.Micro.Mtl (view)
 import Lens.Micro.Internal
 import Lens.Micro.GHC ()
 #else
 import Control.Lens
 #endif
 import Data.Function (on)
-import Data.List (deleteBy)
+import Data.List as L (deleteBy, find)
+import Data.Maybe (fromMaybe)
 import Data.Text
 import Data.Tuple (swap)
 import GHC.Generics (Generic)
+import Codec.Xlsx.Types.SheetState as SheetState (SheetState(..))
+import Prelude hiding (lookup)
 
-newtype SheetList = SheetList{ unSheetList :: [(Text, Worksheet)] }
+newtype SheetList = SheetList{ unSheetList :: [(Text, SheetState, Worksheet)] }
     deriving (Eq, Show, Generic)
 
-type instance IxValue (SheetList) = Worksheet
+type instance IxValue (SheetList) = (SheetState, Worksheet)
 type instance Index (SheetList) = Text
 
 instance Ixed SheetList where
     ix k f sl@(SheetList l) = case lookup k l of
-        Just v  -> f v <&> \v' -> SheetList (upsert k v' l)
+        Just v2  -> f v2 <&> \v2' -> SheetList (upsert k v2' l)
         Nothing -> pure sl
     {-# INLINE ix #-}
 
 instance At SheetList where
-  at k f (SheetList l) = f mv <&> \r -> case r of
-      Nothing -> SheetList $ maybe l (\v -> deleteBy ((==) `on` fst) (k,v) l) mv
-      Just v' -> SheetList $ upsert k v' l
+  at k f (SheetList l) = f mv2 <&> \r -> case r of
+      Nothing -> SheetList $ maybe l (\v2 -> deleteBy ((==) `on` view _1) (t2to3 k v2) l) mv2
+      Just v2' -> SheetList $ upsert k v2' l
     where
-      mv = lookup k l
+      mv2 = lookup k l
   {-# INLINE at #-}
 
-upsert :: (Eq k) => k -> v -> [(k,v)] -> [(k,v)]
-upsert k v [] = [(k,v)]
-upsert k v ((k1,v1):r) =
-    if k == k1
-    then (k,v):r
-    else (k1,v1):upsert k v r
+t2to3 :: k -> (v, v') -> (k, v, v')
+t2to3 k = uncurry (k,,)
+{-# INLINE t2to3 #-}
 
--- | lens giving access to a worksheet from 'Xlsx' object
--- by its name
+t3to2 :: (k, v, v') -> (v, v')
+t3to2 (_, v, v') = (v, v')
+{-# INLINE t3to2 #-}
+
+lookup :: (Eq k) => k -> [(k, v, v')] -> Maybe (v, v')
+lookup k l = L.find ((k ==) . view _1) l <&> t3to2
+
+upsert :: (Eq k) => k -> (v, v') -> [(k, v, v')] -> [(k, v, v')]
+upsert k v2 [] = [t2to3 k v2]
+upsert k v2 (t3@(view _1 -> k'):r) =
```

Comment:
view pattern here just makes code less readable imo. `(t@(k1,_,_))` is more explicit